File: pitch_detection.py
"""
Created for EECS 351 Final Project - Music Transcriber

Authors: Jacob Avery, Ethan Regan, Jae Un Pae

Description: Function that runs a MATLAB script using the "pitch" function.
"""


# Runs Matlab "pitch" function - used on stemss
import matlab.engine
import matplotlib.pyplot as plt


m_eng = matlab.engine.start_matlab()


def MATLAB_Pitches(filename: str, showPlot: bool = False):
    """
    MATLAB script that detects the pitches in an wave file.

    :param filename: path to file
    :param showPlot: plot the pitches
    :return: vector of pitches in Hz, position of pitch in seconds
    """
    pitches, s = m_eng.pitch_detection(filename, nargout=2)

    if showPlot:
        plt.plot(pitches)
        plt.show()
    return pitches, s


if __name__ == '__main__':
    """
    Example use case with the test.wav file    
    """

    MATLAB_Pitches("test.wav", True)

File: pitch_detection.m
%{
Created for EECS 351 Final Project - Music Transcriber

Authors: Jacob Avery, Ethan Regan, Jae Un Pae

Description: Detects pitches from music with mono audio

Requires: Audio Toolbox

Reference:
https://matlab.mathworks.com/

%}

% Filename path relative to project path
function [pitches, s] = pitch_detection(filename)
    [song,fs] = audioread(filename);


    %sound(song,fs)
    
    %Makes the stereo audio mono
    lv = ~sum(song == 0, 2);                       
    mono = sum(song, 2);                             
    mono(lv) = mono(lv)/2;
    
    song = mono;
    
    %power of the song
    %remove the lower power areas to reduce noise
    song_ones = [abs(song) > 0.01];
    song = [song_ones .* song];
    
    %Process the song
    method = "SRH";
    range = [1, 800]; % hertz
    winDur = 0.1; % seconds, increase for lower res
    overlapDur = 0.06; % seconds
    medFiltLength = 10; % frames
    
    winLength = round(winDur*fs);
    overlapLength = round(overlapDur*fs);
    
    [pitches, s] = pitch(song,fs, ...
        Method=method, ...
        Range=range, ...
        WindowLength=winLength, ...
        OverlapLength=overlapLength, ...
        MedianFilterLength=medFiltLength);

    s = s/fs;
    
    % Remove stray notes
    for i = 3:length(pitches) - 2
        left2 = pitches(i-2);
        left1 = pitches(i-1);
        right1 = pitches(i+1);
        right2 = pitches(i+2);
        meanleft = (left2 + left1) / 2;
        meanright = (right2 + right1) / 2;
        if (abs(pitches(i) - meanleft) > 4 && abs(pitches(i) - meanright) > 4)
            pitches(i) = pitches(i-1);
        end
    end
end








